### PR TITLE
chore: add interceptors document

### PR DESCRIPTION
**Pull request by sr-remsha**

### CODE REVIEW

File: docs/tutorials/interceptors.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+* **Generic interceptors** that modify both the incoming request and the response from the upstream (e.g. caching the responses)
+
+Interceptors in AI DIAL can:
+
+* Modify an incoming AI DIAL request received from the client (or it may leave it as is).
+* Call an upstream AI DIAL application (the upstream for short) with the modified request.
+* Modify the response from the upstream (or it may leave it as is).
+* Return the modified response to the client.
+
+For example, interceptors can block requests that violate specific regulations, related with restricted domains, or potentially lead to data leaks or biased responses. Another use case is when interceptors allow applications or models to respond solely to specific subjects and anonymize Personally Identifiable Information (PII) from user requests, or cache LLM responses.
```
Comment (adubovik) [inline docs/tutorials/interceptors.md:20]: `related to restricted domains`

File: docs/tutorials/interceptors.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+Client -> (original request) ->
+  Interceptor 1 -> (modified request #1) ->
+    Interceptor 2 -> (modified request #2) ->
+      Upstream -> (original response) ->
+    Interceptor 2 -> (modified response #1) ->
+  Interceptor 1 -> (modified response #2) ->
+Client
+```
+
+AI DIAL Core manages chat completion requests from interceptors through the endpoint: `/openai/deployments/<deployment|interceptor>/chat/completions`. It uses the deployment name `interceptor` to handle requests from all interceptors. Upon receiving a request, it identifies the next interceptor based on the current interceptor specified in the API Key data. The final interceptor in the sequence is always the target deployment (application, model).
```
Comment (adubovik) [inline docs/tutorials/interceptors.md:40]: through the endpoint: `/openai/deployments/interceptor/chat/completions`. It uses the reserved deployment name `interceptor` to handle requests

Comment (adubovik) [inline docs/tutorials/interceptors.md:40]: Upon receiving a request, it identifies the next interceptor based on its per-request API key.

File: docs/tutorials/interceptors.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+To demonstrate the flow on a more specific example, lets take two interceptors **gtp-cache** and **pii-anonymizer** configured for the **GPT-4** model:
```
Comment (adubovik) [inline docs/tutorials/interceptors.md:42]: typo: gpt-cache

File: docs/tutorials/interceptors.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+```json
+//configuration in AI DIAL Core
```
Comment (adubovik) [inline docs/tutorials/interceptors.md:45]: Here it's called `configuration in AI DIAL Core`, later the **same** config is called `AI DIAL Core dynamic settings`. Let's make it consistent.

File: docs/tutorials/interceptors.md (hunk 1 above)
Comment (sr-remsha) [reply docs/tutorials/interceptors.md]: fixed

File: docs/tutorials/interceptors.md (hunk 2 above)
Comment (sr-remsha) [reply docs/tutorials/interceptors.md]: fixed

Comment (sr-remsha) [reply docs/tutorials/interceptors.md]: fixed

File: docs/tutorials/interceptors.md (hunk 3 above)
Comment (sr-remsha) [reply docs/tutorials/interceptors.md]: fixed

File: docs/tutorials/interceptors.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        "pii-anonymizer": {
+            "endpoint": "${INTERCEPTOR_SERVICE_URL}/openai/deployments/pii-anonymizer/chat/completions",
+            "description": "description"
+        }
+    }
+    ...
+```
+
+### Assign
```
Comment (adubovik) [inline docs/tutorials/interceptors.md:84]: The assign step is duplicated twice, which is confusing.

File: docs/tutorials/interceptors.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        ...
+    },
+    ...
+    }
+```
+
+**Note**: make sure that chat completion interceptors are only used in chat models or application, embeddings interceptors are only used in embeddings models.
+
+## Interceptor Endpoint Specification
```
Comment (adubovik) [inline docs/tutorials/interceptors.md:111]: This section is redundant since it repeats which already has been said or demonstrated above.

Comment (sr-remsha) [reply docs/tutorials/interceptors.md]: fixed

File: docs/tutorials/interceptors.md (hunk 4 above)
Comment (sr-remsha) [reply docs/tutorials/interceptors.md]: fixed

File: docs/tutorials/interceptors.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+For example, to implement PII (Personally Identifiable Information) anonymization for all data sent to models through AI DIAL, you can use a **generic** interceptor which can employ specific locally deployed NLP models to obfuscate (replace with token) PII in requests (pre-interceptor) and decode it in responses (post-interceptor), effectively ensuring the anonymization of all personal data.
+
+<!-- Interceptors in AI DIAL can:
+
+* Modify an incoming AI DIAL request received from the client (or it may leave it as is).
+* Call an upstream AI DIAL application (the upstream for short) with the modified request.
+* Modify the response from the upstream (or it may leave it as is).
+* Return the modified response to the client. -->
+
+The stack of interceptors in AI DIAL works similarly to a stack of middleware in Express.js or Django. Every request/response in the diagram above goes through AI DIAL Core (this is hidden from the diagram for brevity):
```
Comment (astsiapanay) [inline docs/tutorials/interceptors.md:26]: Every request/response in the diagram **above** -> **below** goes through AI DIAL Core (this is hidden from the diagram for brevity):

Comment (astsiapanay) [inline docs/tutorials/interceptors.md:26]: The sentence refers to the stacks but doesn't contains any links to examples. That may confuse a reader. I would suggest adding links to the stacks or re-phrase the sentence not mentioning any stacks.

Comment (sr-remsha) [reply docs/tutorials/interceptors.md]: fixed

**Pull request merged**